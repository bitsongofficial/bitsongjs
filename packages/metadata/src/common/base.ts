import { z } from 'zod';
import { BaseMetadataAttribute, BaseMetadataAttributeSchema } from "./attribute";
import { Markdown, URI, markdown, nonEmptyStringSchema, uriSchema } from "./primitive";
import { RecursiveUnbrand } from '../utils';

/**
 * BaseMetadata is the most used standard for nfts, used by opensea and other marketplaces
 * 
 * @see https://docs.opensea.io/docs/metadata-standards#metadata-structure
 */
export type BaseMetadata = {
  /**
   * Name of the NFT.
   */
  name?: string;
  /**
   * A human-readable description of the item. Markdown is supported.
   */
  description?: Markdown;
  /**
   * This is the URL to the image of the item. Can be just about any type of image, IPFS or Arweave URLs or paths
   */
  image?: URI;
  /**
   * A URL to a multi-media attachment for the item. The file extensions GLTF, GLB, WEBM, MP4, M4V, OGV, and OGG are supported, along with the audio-only extensions MP3, WAV, and OGA.
   */
  animation_url?: URI;
  /**
   * This is the URL that will appear below the asset's image on OpenSea and will allow users to leave OpenSea and view the item on your site.
   */
  external_url?: URI;
  /**
   * Attributes are traits that can be used to sort and filter items on Marketplaces. They are also displayed on the item page.
   */
  attributes?: BaseMetadataAttribute[];
}

export const BaseMetadataSchema = z.object({
  description: markdown(
    z.string({
      description: 'A human-readable description of the item. Markdown is supported.',
    }),
  ).optional(),

  external_url: uriSchema(
    'This is the URL to the image of the item. Can be just about any type of image, IPFS or Arweave URLs or paths.',
  ).optional(),

  name: z.string({ description: 'Name of the NFT.' }).optional(),

  attributes: BaseMetadataAttributeSchema.array()
    .optional()
    .describe(
      'Attributes are traits that can be used to sort and filter items on Marketplaces. They are also displayed on the item page.',
    ),

  image: uriSchema('Marketplaces will store any NFT image here.').optional(),

  animation_url: uriSchema(
    'A URL to a multi-media attachment for the item. The file extensions GLTF, GLB, WEBM, MP4, M4V, OGV, ' +
    'and OGG are supported, along with the audio-only extensions MP3, WAV, and OGA.',
  ).optional(),
});

export type ContentMetadataCommon = {
  id: string;
}

const ContentMetadataCommonSchema = z.object({
  id: nonEmptyStringSchema(
    'A unique identifier that in storages like IPFS ensures the uniqueness of the metadata URI. Use a UUID if unsure.',
  ),
});

export function metadataDetailsWith<Augmentation extends {}>(augmentation: Augmentation) {
  return ContentMetadataCommonSchema.extend(augmentation);
}

export function contentWith<Augmentation extends {
  $schema: z.ZodLiteral<string>;
  bitsong: ReturnType<typeof metadataDetailsWith>;
}>(augmentation: Augmentation) {
  return BaseMetadataSchema.extend(augmentation);
}

export type BaseDetails = RecursiveUnbrand<BaseMetadata>;