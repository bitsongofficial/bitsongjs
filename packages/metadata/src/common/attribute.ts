import { z } from 'zod';
import { nonEmptyStringSchema } from './primitive';

/**
 * The display type of a marketplace metadata attribute.
 *
 * @see https://docs.opensea.io/docs/metadata-standards#attributes
 */
export enum BaseMetadataAttributeDisplayType {
  NUMBER = 'number',
  STRING = 'string',
  DATE = 'date',
}

/**
 * Base metadata attribute.
 * 
 * @see https://docs.opensea.io/docs/metadata-standards#attributes
 */
export type BaseMetadataAttribute = {
  /**
   * The trait type.
   */
  trait_type?: string | undefined;
  /**
   * The trait value.
   */
  value?: string | number | undefined;
  /**
   * The trait display type.
   */
  display_type?: BaseMetadataAttributeDisplayType | undefined;
};

export const BaseMetadataAttributeSchema: z.ZodType<BaseMetadataAttribute, z.ZodTypeDef, object> = z
  .object({
    trait_type: nonEmptyStringSchema('The name of the trait.').optional(),
    value: z.union([z.string(), z.number()]).optional(),
    display_type: z.nativeEnum(BaseMetadataAttributeDisplayType).optional(),
  })
  .passthrough();