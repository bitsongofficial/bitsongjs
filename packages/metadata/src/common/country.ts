import iso from 'iso-3166-1'
import { z } from 'zod';
import { Brand } from '../utils';

export const CountryType = iso.all()
  .map((country) => {
    return {
      id: country.alpha2,
      name: country.country,
    }
  })
  .sort((a, b) => a.name.localeCompare(b.name))

export type Country = Brand<string, 'Country'>;

export function toCountry(value: string): Country {
  const lowerCaseValue = value.toLowerCase();
  const country = CountryType.find(country => country.id.toLowerCase() === lowerCaseValue || country.name.toLowerCase() === lowerCaseValue);
  return country ? (country.id as Country) : (value as Country);
}

export function countrySchema(
  description: string = 'The country of the track',
): z.ZodType<Country, z.ZodTypeDef, unknown> {
  return z
    .string({ description })
    .refine(value => CountryType.some(country => country.id === value || country.name === value), {
      message: 'Invalid country'
    })
    .transform(toCountry);
}