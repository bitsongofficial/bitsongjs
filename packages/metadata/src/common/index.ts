export * from './attribute'
export * from './base'
export * from './mime'
export * from './primitive'