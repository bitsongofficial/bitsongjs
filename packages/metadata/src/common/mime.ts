export enum MediaAudioMimeType {
  WAV = 'audio/wav',
  WAV_VND = 'audio/vnd.wave',
  MP3 = 'audio/mpeg',
  OGG_AUDIO = 'audio/ogg',
  MP4_AUDIO = 'audio/mp4',
  AAC = 'audio/aac',
  WEBM_AUDIO = 'audio/webm',
  FLAC = 'audio/flac',
}