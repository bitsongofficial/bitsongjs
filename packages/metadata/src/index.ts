export * from './artist';
export * from './author_publisher';
export * from './common'
export * from './explicit';
export * from './genre';
export * from './license';
export * from './schema';
export * from './track';
export * from './utils';