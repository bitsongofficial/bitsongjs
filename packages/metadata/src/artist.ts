import { z } from "zod";
import { BitsongAddress, bitsongAddressSchema } from "./common";
import { Brand } from "./utils";

export const artistRoles = [{
  id: 'main_artist',
  name: 'Main Artist',
  description: 'The main artist of the track.',
}, {
  id: 'featuring',
  name: 'Featuring',
  description: 'The artist is featured in the track.',
}, {
  id: 'remixed_by',
  name: 'Remixed By',
  description: 'The artist remixed the track.',
}, {
  id: 'versus',
  name: 'Versus (vs)',
  description: 'The artist is against another artist.',
}, {
  id: 'with',
  name: 'With',
  description: 'The artist is with another artist.',
}];

export type ArtistRole = Brand<string, 'ArtistRole'>;

export function toArtistRole(value: string): ArtistRole {
  const role = artistRoles.find(role => role.id === value || role.name === value);
  return role ? (role.name as ArtistRole) : (value as ArtistRole);
}

export function artistRoleSchema(
  description: string = 'The role of the artist',
): z.ZodType<ArtistRole, z.ZodTypeDef, unknown> {
  return z
    .string({ description })
    .refine(value => artistRoles.some(role => role.id === value || role.name === value), {
      message: 'Invalid artist role'
    })
    .transform(toArtistRole);
}

export type ArtistDetails = {
  address?: BitsongAddress;
  name: string;
  role: ArtistRole;
}

export const ArtistDetailsSchema = z.object({
  address: bitsongAddressSchema().optional(),
  name: z.string({ description: 'The name of the artist.' }),
  role: artistRoleSchema('The role of the artist.'),
});