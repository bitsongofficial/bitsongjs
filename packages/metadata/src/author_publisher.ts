import { z } from "zod";
import { BitsongAddress, bitsongAddressSchema } from "./common";
import { Brand } from "./utils";

export const authorPublisherRoles = [{
  id: 'arranger',
  name: 'Arranger',
  description: 'The artist arranged the track.',
}, {
  id: 'composer',
  name: 'Composer',
  description: 'The artist composed the track.',
}, {
  id: 'lyricist',
  name: 'Lyricist',
  description: 'The artist wrote the lyrics of the track.',
}, {
  id: 'producer',
  name: 'Producer',
  description: 'The artist produced the track.',
}, {
  id: 'publisher',
  name: 'Publisher',
  description: 'The artist published the track.',
}, {
  id: 'songwriter',
  name: 'Songwriter',
  description: 'The artist wrote the song.',
}];

export type AuthorPublisherRole = Brand<string, 'AuthorPublisherRole'>;

export function toAuthorPublisherRole(value: string): AuthorPublisherRole {
  const role = authorPublisherRoles.find(role => role.id === value || role.name === value);
  return role ? (role.name as AuthorPublisherRole) : (value as AuthorPublisherRole);
}

export function authorPublisherRoleSchema(
  description: string = 'The role of the author or publisher',
): z.ZodType<AuthorPublisherRole, z.ZodTypeDef, unknown> {
  return z
    .string({ description })
    .refine(value => authorPublisherRoles.some(role => role.id === value || role.name === value), {
      message: 'Invalid author or publisher role'
    })
    .transform(toAuthorPublisherRole);
}

export type AuthorPublisherDetails = {
  address?: BitsongAddress;
  name: string;
  role: AuthorPublisherRole;
}

export const AuthorPublisherDetailsSchema = z.object({
  address: bitsongAddressSchema().optional(),
  name: z.string({ description: 'The name of the author or publisher.' }),
  role: authorPublisherRoleSchema('The role of the author or publisher.'),
});