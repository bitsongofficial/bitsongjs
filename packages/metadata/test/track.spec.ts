import { toLicense, trackMetadata, toTrackGenre, toExplicit, toArtistRole } from '../src'
import { toMarkdown, toCountry } from '../src/common'

describe('Given the track metadata', () => {
  describe('when using Track Metadata', () => {
    it('should return a valid track metadata', () => {
      const metadata = trackMetadata({
        title: 'title',
        titleLocale: 'en',
        description: toMarkdown(`this is a description`),
        artists: [
          {
            name: 'The best Artist',
            role: toArtistRole('Main Artist'),
          },
          {
            name: 'Artist 2',
            role: toArtistRole('Featuring'),
          },
        ],
        artwork: 'https://bitsong.io/artwork.png',
        audio: 'https://bitsong.io/audio.mp3',
        duration: 100,
        license: toLicense("No Rights Reserved (CC0)"), // or toLicense('no_rights_reserved')
        genre: toTrackGenre('ambient'),
        country: toCountry('IT'), // or toCountry('it') or toCountry('Italy') or toCountry('italy')
        explicit: toExplicit('Clean'), // or toExplicit('clean') or toExplicit('Explicit') or toExplicit('explicit')
        isrc: 'IT-ABC-12345',
        authors_publishers: [
          {
            name: 'Author 1',
            role: toArtistRole('Composer'),
          },
          {
            name: 'Author 2',
            role: toArtistRole('Lyricist'),
          },
        ],
        cLine: '2020 The Best Label',
        pLine: '2020 The Best Label',
        liveRecording: false,
        lyrics: toMarkdown(`this is a lyrics`),
        lyricsLocale: 'en',
        previousRelease: false,
        previewStartTime: 15,
        version: 'Radio Edit',
        video: 'https://bitsong.io/video.mp4',
        iswc: 'IT-ABC-12345',
        label: 'The Best Label',
      })

      expect(metadata).toMatchSnapshot({
        bitsong: {
          id: expect.any(String),
        },
      })
    })
  })
})