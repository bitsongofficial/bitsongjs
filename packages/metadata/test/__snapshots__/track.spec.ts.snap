// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the track metadata when using Track Metadata should return a valid track metadata 1`] = `
Object {
  "$schema": "https://json-schemas.bitsong.io/metadata/track/1.0.0.json",
  "animation_url": "https://bitsong.io/video.mp4",
  "attributes": Array [
    Object {
      "display_type": "string",
      "trait_type": "License",
      "value": "No Rights Reserved (CC0)",
    },
    Object {
      "display_type": "number",
      "trait_type": "Duration",
      "value": 100,
    },
    Object {
      "display_type": "string",
      "trait_type": "Artists",
      "value": "The best Artist, Artist 2",
    },
    Object {
      "display_type": "string",
      "trait_type": "Genre",
      "value": "Ambient",
    },
    Object {
      "display_type": "string",
      "trait_type": "Explicit",
      "value": "Clean",
    },
    Object {
      "display_type": "string",
      "trait_type": "Version",
      "value": "Radio Edit",
    },
    Object {
      "display_type": "string",
      "trait_type": "P-Line",
      "value": "2020 The Best Label",
    },
    Object {
      "display_type": "string",
      "trait_type": "C-Line",
      "value": "2020 The Best Label",
    },
    Object {
      "display_type": "string",
      "trait_type": "Country",
      "value": "IT",
    },
    Object {
      "display_type": "string",
      "trait_type": "ISRC",
      "value": "IT-ABC-12345",
    },
    Object {
      "display_type": "string",
      "trait_type": "ISWC",
      "value": "IT-ABC-12345",
    },
    Object {
      "display_type": "string",
      "trait_type": "Label",
      "value": "The Best Label",
    },
    Object {
      "display_type": "string",
      "trait_type": "Authors & Publishers",
      "value": "Author 1, Author 2",
    },
  ],
  "bitsong": Object {
    "artists": Array [
      Object {
        "name": "The best Artist",
        "role": "Main Artist",
      },
      Object {
        "name": "Artist 2",
        "role": "Featuring",
      },
    ],
    "artwork": "https://bitsong.io/artwork.png",
    "audio": "https://bitsong.io/audio.mp3",
    "authors_publishers": Array [
      Object {
        "name": "Author 1",
        "role": "Composer",
      },
      Object {
        "name": "Author 2",
        "role": "Lyricist",
      },
    ],
    "cLine": "2020 The Best Label",
    "country": "IT",
    "description": "this is a description",
    "duration": 100,
    "explicit": "Clean",
    "genre": "Ambient",
    "id": Any<String>,
    "isrc": "IT-ABC-12345",
    "iswc": "IT-ABC-12345",
    "label": "The Best Label",
    "license": "No Rights Reserved (CC0)",
    "liveRecording": false,
    "lyrics": "this is a lyrics",
    "lyricsLocale": "en",
    "pLine": "2020 The Best Label",
    "previewStartTime": 15,
    "previousRelease": false,
    "title": "title",
    "titleLocale": "en",
    "version": "Radio Edit",
    "video": "https://bitsong.io/video.mp4",
  },
  "description": "this is a description",
  "image": "https://bitsong.io/artwork.png",
  "name": "title",
}
`;
